src/library/scala/collection/immutable/ArraySeq.scala

```diff
@@ -208,18 +208,21 @@ object ArraySeq extends StrictOptimizedClassTagSeqFactory[ArraySeq] { self =>
 
   def empty[A : ClassTag]: ArraySeq[A] = emptyImpl
 
-  def from[A : ClassTag](it: scala.collection.IterableOnce[A]): ArraySeq[A] = unsafeWrapArray {
-    val n = it.knownSize
-    if (n > -1) {
-      val elements = Array.ofDim[A](n)
-      val iterator = it.iterator
-      var i = 0
-      while (i < n) {
-        ScalaRunTime.array_update(elements, i, iterator.next())
-        i = i + 1
-      }
-      elements
-    } else ArrayBuffer.from(it).toArray
+  def from[A](it: scala.collection.IterableOnce[A])(implicit tag: ClassTag[A]): ArraySeq[A] = it match {
+    case as: ArraySeq[A] if as.elemTag == tag => as
+    case _ => unsafeWrapArray {
+      val n = it.knownSize
+      if (n > -1) {
+        val elements = Array.ofDim[A](n)
+        val iterator = it.iterator
+        var i = 0
+        while (i < n) {
+          ScalaRunTime.array_update(elements, i, iterator.next())
```

Comment:
not clear to me why this call is manual rather than letting the compiler insert it